src/global/a11y/interactivity-checker.ts

```diff
@@ -0,0 +1,38 @@
+export function hasGeometry(element: HTMLElement): boolean {
+  // Use logic from jQuery to check for an invisible element.
+  // See https://github.com/jquery/jquery/blob/master/src/css/hiddenVisibleSelectors.js#L12
+  return !!(
+    element.offsetWidth ||
+    element.offsetHeight ||
+    (typeof element.getClientRects === 'function' && element.getClientRects().length)
+  );
+}
+
+export interface InterfaceInteractivityChecker {
+  isVisible(element: HTMLElement): boolean;
+}
+
+export class InteractivityChecker implements InterfaceInteractivityChecker {
+  /**
+   * Gets whether an element is visible for the purposes of interactivity.
+   *
+   * This will capture states like `display: none` and `visibility: hidden`, but not things like
+   * being clipped by an `overflow: hidden` parent or being outside the viewport.
+   *
+   * @returns Whether the element is visible.
+   */
+  public isVisible(element: HTMLElement): boolean {
+    return hasGeometry(element) && getComputedStyle(element).visibility === 'visible';
+  }
+}
+
+class InteractivityCheckerMock implements InterfaceInteractivityChecker {
```

Comment:
I was surprised to see that the check works for the e2e tests, which is good. At this point I wonder if the Mock function is still necessary. Could it be removed?

Reply:
The mock is needed for the unit tests.